### PR TITLE
SELC-2494 Add categories filter for search AOO and UO

### PR DESCRIPTION
**Pull request by flaminiaScarciofolo**

### CODE REVIEW

File: core/src/main/java/it/pagopa/selfcare/party/registry_proxy/core/AOOService.java [hunk 1]
```diff
@@ -3,11 +3,12 @@
...
 
+import java.util.List;
+
 
 public interface AOOService {
 
     QueryResult<AOO> findAll(int page, int limit);
-
-    AOO findByUnicode(String codiceUniAOO);
+    AOO findByUnicode(String codiceUniAOO, List<String> categoriesList);
```
Comment (manuraf) [inline core/src/main/java/it/pagopa/selfcare/party/registry_proxy/core/AOOService.java:12]: should be named `findByUnicodeAndCategories`

Comment (flaminiaScarciofolo) [reply core/src/main/java/it/pagopa/selfcare/party/registry_proxy/core/AOOService.java:12]: il filtro per categoria non è obbligatorio. cambiamo comunque il nome?

File: core/src/main/java/it/pagopa/selfcare/party/registry_proxy/core/InstitutionServiceImpl.java [hunk 2]
```diff
@@ -66,15 +66,10 @@ public Institution findById(String id, Optional<Origin> origin, List<String> cat
...
-                    .orElseGet(categories.size() > 0 ? new Supplier<List<Institution>>() {
-                        @Override
-                        public List<Institution> get() {
-                            return new ArrayList<>();
-                        }
-                    } : institutionsSupplier);
+                                    (categories.isEmpty() || categories.contains(institution.getCategory()))
+                            )
+                            .collect(Collectors.toList()))
+                    .orElseGet(!categories.isEmpty() ? ArrayList::new : institutionsSupplier);
```
Comment (manuraf) [inline core/src/main/java/it/pagopa/selfcare/party/registry_proxy/core/InstitutionServiceImpl.java:72]: could be more readable something like this
```
final Supplier<List<Institution>> institutionsSupplier = () -> indexSearchService.findById(Field.ID, id);
List<Institution> institutions;

if (origin.isPresent()) {
    Origin orig = origin.get();
    institutions = institutionsSupplier.get().stream()
            .filter(institution -> institution.getOrigin().equals(orig) &&
                    (categories.isEmpty() || categories.contains(institution.getCategory())))
            .collect(Collectors.toList());
} else {
    institutions = categories.isEmpty() ? new ArrayList<>() : institutionsSupplier.get();
}
```

**Review by manuraf — CHANGES_REQUESTED** [review]

File: core/src/main/java/it/pagopa/selfcare/party/registry_proxy/core/AOOService.java (hunk 1 above)
Comment (manuraf) [reply core/src/main/java/it/pagopa/selfcare/party/registry_proxy/core/AOOService.java:12]: that's ok, let's leave it like this

File: web/src/main/java/it/pagopa/selfcare/party/registry_proxy/web/controller/AOOController.java [hunk 3]
```diff
@@ -58,10 +60,17 @@ public AOOsResource findAll(@ApiParam(value = "${swagger.model.*.page}")
...
+                                     @RequestParam(value = "categories", required = false)
+                                     String categories) {
         log.trace("find AOO start");
         log.debug("find AOO codiceUniAoo = {}", codiceUniAoo);
-        final AOOResource aooResource = AOOMapper.toResource(aooService.findByUnicode(codiceUniAoo));
+        List<String> categoriesList = Collections.emptyList();
+        if (StringUtils.hasText(categories)) {
+            categoriesList = Arrays.stream(categories.split(",")).collect(Collectors.toList());
+        }
+        final AOOResource aooResource = AOOMapper.toResource(aooService.findByUnicode(codiceUniAoo,categoriesList));
```
Comment (manuraf) [inline web/src/main/java/it/pagopa/selfcare/party/registry_proxy/web/controller/AOOController.java:73]: look at https://medium.com/@AADota/spring-passing-list-and-array-of-values-as-url-parameters-1ed9bbdf0cb2

File: web/src/main/java/it/pagopa/selfcare/party/registry_proxy/web/controller/UOController.java [hunk 4]
```diff
@@ -58,11 +60,18 @@ public UOsResource findAll(@ApiParam(value = "${swagger.model.*.page}")
...
+                                    @ApiParam(value = "${swagger.model.*.categories}")
+                                    @RequestParam(value = "categories", required = false)
+                                    String categories) {
         log.trace("find UO start");
         log.debug("find UO = {}", codiceUniUo);
-        final UOResource uoResource = UOMapper.toResource(uoService.findByUnicode(codiceUniUo));
+        List<String> categoriesList = Collections.emptyList();
+        if (StringUtils.hasText(categories)) {
+            categoriesList = Arrays.stream(categories.split(",")).collect(Collectors.toList());
+        }
```
Comment (manuraf) [inline web/src/main/java/it/pagopa/selfcare/party/registry_proxy/web/controller/UOController.java:73]: ditto

**Review by manuraf — CHANGES_REQUESTED** [review]

**Review by manuraf — APPROVED** [review]

**Pull request merged**